### PR TITLE
kokkos-tools: new package

### LINKED ISSUE #45149: Request To Add Kokkos-tools to spack package list
### Summary

Would be appreciated if kokkos-tools can be added to spack’ list of packages. 

### Rationale

_No response_

### Description

Installing kokkos-tools separate of spack defeats the purpose of using spack. Not to mention installing separate of spack can lead to unintentional conflicting dependencies being installed. 

### Additional information

_No response_

### General information

- [X] I have searched the issues of this repo and believe this is not a duplicate

### PR DESCRIPTION
**Pull request by rbberger**
Adds a basic spackage for kokkos-tools

I've deliberately didn't include any of the releases, since the last one is many years old.

The CMake configuration will primarily be dependent on how kokkos was configured and built. So I've only added the `mpi` and `papi` variant for now to create the base skeleton for this.

Basic usage after install:
```bash
export KOKKOS_TOOLS_LIBS=$(spack location -i kokkos-tools)/lib/<name_of_tool_shared_lib>.so
```

Closes #45149 

FYI @dhope-lab @lucbv

### CODE REVIEW

**Review by tldahlgren — APPROVED** [review]
Confirmed homepage and branch, rest LGTM.

**Comment by tldahlgren** [comment]
Will this be ready to merge once checks pass or should I wait for @lucbv and or @dhope-lab?

**Comment by rbberger** [comment]
@tldahlgren I'd wait for some feedback from the Kokkos devs. Maybe there was a reason this didn't exist yet.

**Comment by dhope-lab** [comment]
Hello,

definitely want to take a look at this tonight. I’m interested in having all the dependencies installed from the CMakeList.txt file instead of having a barebones version.

**Comment by vlkale** [comment]
@jennfshr and I were looking at this just a few weeks ago on July 3 and were working to develop it within Sandia - I was explaining the CMake build system.  I need to catch up with her but am wondering if this is a part of that or a coincidence you are working on the same thing? (@jennfshr?)

**Comment by vlkale** [comment]
> @tldahlgren I'd wait for some feedback from the Kokkos devs. Maybe there was a reason this didn't exist yet.

FYI, this hadn't been done yet because the CMake build system for Kokkos Tools is relatively recent (and we needed to focus on maintaining updating that) and also Kokkos-core documentation/viewpoint for building has not encouraged Spack and prefers CMake for building in the last couple years (maybe this has changed recently in the last few months). However, I believe that as Kokkos Tools is starting to use newer tpls and connectors, e.g., Apex, ScoreP, a Spack package for Kokkos Tools becomes even more important and high-priority.

**Comment by rbberger** [comment]
@vlkale no this is an independent effort. born out of the Venado hackathon we just had at LANL. This skeleton can easily be extended to enable more variants/features. This is just meant to get the ball rolling.

**Comment by vlkale** [comment]
> @vlkale no this is an independent effort. born out of the Venado hackathon we just had at LANL. 

Got it, thanks for clarifying that for me. 

> This skeleton can easily be extended to enable more variants/features. This is just meant to get the ball rolling.

Sounds good. We can use this PR to develop and I will contribute+review along with @jennfshr. I (and @jennfshr, if she would like) would like to be designated the official maintainer of the Kokkos Tools Spack package, if that is OK.

**Comment by vlkale** [comment]
> Will this be ready to merge once checks pass or should I wait for @lucbv and or @dhope-lab?

I have looked at this but I would like a little more time towards at least end of this week, if that's OK, to double-check with other Kokkos Tools devs. 

There is certainly more to add but this is a good starting point that can be merged.

**Comment by vlkale** [comment]
Hi all, We have had some Sandia folks that are on vacation, and we need till next Monday morning to review, if you don’t mind. Would that work?

**Comment by dhope-lab** [comment]
> Hi all, We have had some Sandia folks that are on vacation, and we need till next Monday morning to review, if you don’t mind. Would that work?

Fine by me.

**Comment by tldahlgren** [comment]
Ping me when you're ready for this PR to be merged.

**Comment by vlkale** [comment]
> Ping me when you're ready for this PR to be merged.

Will do. Still a bit delayed but aiming for early this week.

Actually, I think if either @crtrott or @dalg24 are good with this, then we can merge.

File: var/spack/repos/builtin/packages/kokkos-tools/package.py
Comment (dalg24) [inline var/spack/repos/builtin/packages/kokkos-tools/package.py:22]: Kokkos is NOT a required dependency to build the tools.
You only need it to run the tests/examples.

Comment (rbberger) [reply var/spack/repos/builtin/packages/kokkos-tools/package.py:22]: It was easier to just add the kokkos dependency since things like nvtx connector require `Kokkos_ENABLE_CUDA`. You can make it independent and just set that variable via CMake, but that solution seemed more messy and would create duplication in dependency management.

Comment (dalg24) [reply var/spack/repos/builtin/packages/kokkos-tools/package.py:22]: This dependency on a Kokkos setting for the nvtx and roctx connectors is just silly.  I will make sure it is enabled if CUDA/ROCm are detected.

Comment (rbberger) [reply var/spack/repos/builtin/packages/kokkos-tools/package.py:22]: Ping me once it's fixed. I'll update the Spack package.

**Comment by tldahlgren** [comment]
> > This skeleton can easily be extended to enable more variants/features. This is just meant to get the ball rolling.
> 
> Sounds good. We can use this PR to develop and I will contribute+review along with @jennfshr. I (and @jennfshr, if she would like) would like to be designated the official maintainer of the Kokkos Tools Spack package, if that is OK.

FWIW IMHO I think that's a *great* idea.

**Comment by vlkale** [comment]
@rbberger I have reviewed and this looks good to me, with one exception: there should be a comment/disclaimer in this Spack package that this works only for tools of Kokkos Tools not needing external libraries except PAPI and MPI. 

To elaborate:

1. This Spack package should not make promises right now for supporting nvtx-connector / roctx-connector / vtune-connector. Once Kokkos Tools Github Issue 268 (github.com/kokkos/kokkos-tools/issues/268) is resolved, we can continue with guaranteeing support for these vendor tool connectors of Kokkos Tools.

2. The Kokkos Tools Spack package should be documented to be immediately useful for tool libraries with no library dependences except possibly MPI, e.g., kernel-logger, simple-kernel-timer, space-time-stack, as well as the papi-connector.

I would like to have @dalg24's support for this direction before proceeding. 

Edit: the above assumes we remove the Kokkos dependence in the Spack package.

File: var/spack/repos/builtin/packages/kokkos-tools/package.py
Comment (vlkale) [reply var/spack/repos/builtin/packages/kokkos-tools/package.py:22]: For an initial version, removing the dependence on Kokkos is fine - but we should also make a note somewhere (perhaps in comments or notes in the Spack package) that we don't provide a guarantee of support of vendor connectors, in particular, nvtx-connector/roctx-connector/vtune-connector.

**Review by vlkale — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/kokkos-tools/package.py
Comment (crtrott) [reply var/spack/repos/builtin/packages/kokkos-tools/package.py:22]: I am ok merging as is, and then fixing when we have fixed the build system on the project side.

**Review by crtrott — APPROVED** [review]

**Pull request merged**
